core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1132,6 +1413,75 @@ public SingleFileStoreConfiguration getConfiguration() {
       return configuration;
    }
 
+   @Override
+   public CompletionStage<Void> addSegments(IntSet segments) {
+      if (!segmented) {
+         // Nothing to do
+         return CompletableFutures.completedNull();
+      }
+      return blockingManager.runBlocking(() -> blockingAddSegments(segments), "sfs-addSegments");
+   }
+
+   private void blockingAddSegments(IntSet segments) {
+      long stamp = resizeLock.writeLock();
+      try {
+         for (int segment : segments) {
+            if (entries[segment] != null)
+               continue;
+
+            // Only use LinkedHashMap (LRU) for entries when cache store is bounded
+            Map<K, FileEntry> entryMap = configuration.maxEntries() > 0 ?
+                                         new LinkedHashMap<>(16, 0.75f, true) :
+                                         new HashMap<>();
+            entries[segment] = Collections.synchronizedMap(entryMap);
+         }
+      } finally {
+         resizeLock.unlockWrite(stamp);
+      }
+   }
+
+   @Override
+   public CompletionStage<Void> removeSegments(IntSet segments) {
+      if (!segmented) {
+         // Delete keys individually
+         return Flowable.fromPublisher(publishKeys(segments, null))
+                        .doOnNext(k -> delete(0, k))
```

Comment:
Do we want to provide some back pressure here to prevent blowing up the blocking pool?

Reply:
I forgot that delete() returns a CompletionStage :)